### DISCUSSION TITLE
Replace the default Phase4PeppolServletMessageProcessorSPI

### DISCUSSION THREAD

**Discussion by MakakWasTaken** [phax/phase4#87] [Q&A] [answered]
I am currently trying to send a message with an xml file as the attachment. The XML file has the element 'Nomination_Document' at the root. This causes some problems with the `Phase4PeppolServletMessageProcessorSPI`.

I have written a Custom handler for the SBD and added it to META-INF/services. It seems like this is not propegated properly, because I am getting the error:
```
2022-05-25 11:05:41.539 ERROR 13435 --- [nio-8080-exec-1] c.h.j.LoggingJAXBReadExceptionHandler    : Failed to parse XML document: cvc-elt.1.a: Cannot find the declaration of element 'Nomination_Document'.
2022-05-25 11:05:41.541 ERROR 13435 --- [nio-8080-exec-1] s.Phase4PeppolServletMessageProcessorSPI : [4e03cf71-e868-493a-9387-62cc92bf39c0@phase4] Peppol SBDH Issue: [fatal_error] @ (2:209) cvc-elt.1.a: Cannot find the declaration of element 'Nomination_Document'. (org.xml.sax.SAXParseException: cvc-elt.1.a: Cannot find the declaration of element 'Nomination_Document'.)
2022-05-25 11:05:41.543  WARN 13435 --- [nio-8080-exec-1] c.h.phase4.servlet.AS4RequestHandler     : Error invoking synchronous SPIs
```

Is it possible to replace the functionality in Phase4PeppolServletMessageProcessorSPI to allow for the custom element declaration?

**Comment by phax** [comment, marked as answer]
For the usage with ENTSOG you cannot use the `phase4-peppol-servlet` or `phase4-peppol-server-webapp` submodule, because it adds an additional "SPI layer" that enforces the usage of the "Standard Business Document Header" (SBDH).
To retrieve "arbitrary" AS4 messages, you need to implement the `com.helger.phase4.servlet.spi.IAS4ServletMessageProcessorSPI` SPI interface that is more generic. An example is shown in the generic `phase4-server-webapp` module.

**Reply by MakakWasTaken** [reply]
I tried using the `ExampleReceiveMessageProcessorSPI` from `phase4-server-webapp` to create the service instead. 

The error remains the same.

**Reply by phax** [reply]
Okay, please remove the Peppol SPI - otherwise both SPIs are invoked. Also remove the `phase4-peppol-servlet` module from `pom.xml` because otherwise the SPI is invoked implicitly.

**Reply by MakakWasTaken** [reply]
I figured out why it kept giving me the error. It was because I was yet to remove 

```
    <dependency>
      <groupId>com.helger.phase4</groupId>
      <artifactId>phase4-peppol-servlet</artifactId>
    </dependency>
```
from my `pom.xml`. This caused the servlet to default to still contain some Peppol functions. 

This fixed the error of removing the Peppol SPI, but it does still not load my new service properly.

I have, as mentioned, added it to the `META-INF/services` folder

Does it also need to be specified in the `pom.xml`?

**Log:**
```
2022-05-25 12:06:25.750  WARN 14053 --- [nio-8080-exec-1] .p.s.m.AS4ServletMessageProcessorManager : No AS4 message processor is registered. All incoming messages will be discarded!
2022-05-25 12:06:25.751 ERROR 14053 --- [nio-8080-exec-1] c.h.phase4.servlet.AS4RequestHandler     : No IAS4ServletMessageProcessorSPI is available to process an incoming message
```


The `AS4ServletMessageProcessorManager` does not seem to have any functions to manually add it. I suspect this is because it is supposed to load from the `META-INF/services` folder

**Reply by phax** [reply]
What a timing ;-)

**Reply by phax** [reply]
You need to create the file `src/main/resources/META-INF/services/com.helger.phase4.servlet.spi.IAS4ServletMessageProcessorSPI` that contains the fully qualified name of your `IAS4ServletMessageProcessorSPI` implementation.

An SPI does not need to be included in the pom.xml. It is resolved purely by its name.

**Reply by MakakWasTaken** [reply]
Ahhh, I had named the file `com.helger.phase4.springboot.spi.ENTSOGReceiveMessageProcessorSPI` instead of `com.helger.phase4.servlet.spi.IAS4ServletMessageProcessorSPI`.
I though the naming should be the same as the SPI I made.

Thanks :)

**Reply by phax** [reply]
Yes, it is the other way around, because the file could contain multiple implementations

**Reply by oaxfreeman** [reply]
Hi guys, 
 
 I'm sorry but I don't get it. I have created @Service based on ExampleRecieveMessageProcessorSPI
 
```
 @IsSPIImplementation
@Service
public class ExampleReceiveMessageProcessorSPI implements IAS4ServletMessageProcessorSPI
{
  private static final Logger LOGGER = LoggerFactory.getLogger (ExampleReceiveMessageProcessorSPI.class);
...
```

Yet it is no being registered anywhere:

`2024-07-10T08:36:41.682+02:00  WARN 34444 --- [nio-8080-exec-3] .p.s.m.AS4ServletMessageProcessorManager : No AS4 message processor is registered. All incoming messages will be discarded!
2024-07-10T08:36:41.683+02:00 ERROR 34444 --- [nio-8080-exec-3] c.h.phase4.servlet.AS4RequestHandler     : No IAS4ServletMessageProcessorSPI is available to process an incoming message`

From your discusson I assume that it should be picked up from registered beans. So how to provide impl of `IAS4ServletMessageProcessorSPI`?

Same goes for manual registration:
![image](https://github.com/phax/phase4/assets/8730277/75c2cc23-5e30-443a-94d3-f6d58145e41f)

**Reply by MakakWasTaken** [reply]
Can you confirm that you have a file in your resources folder with the following naming?
![image](https://github.com/phax/phase4/assets/11789635/db1eb454-90dd-468d-8bf9-9f623f713c0e)
It needs to be named exactly: `com.helger.phase4.servlet.spi.IAS4ServletMessageProcessorSPI` otherwise the service is not going to pick it up, which was the error I had above.

Also the content of your file should be something along with:
`com.project.ExampleReceiveMessageProcessorSPI`
Essentially describing the path of the class. This needs to be exact otherwise the message processor is not going to be picked up.

**Reply by oaxfreeman** [reply]
Thank you for your reply, I have never encountered such pattern. Not happy with it, but will do. Thank you @MakakWasTaken

**Reply by phax** [reply]
@piotr-dajlido this is called SPI (Service Provider Interface) and is described at https://docs.oracle.com/javase/tutorial/sound/SPI-intro.html - it's not a Spring thing, but a Java thing :)
